includes/class-wpcom-legacy-redirector.php

```diff
@@ -69,16 +69,17 @@ static function maybe_do_redirect() {
 			return;
 		}
 
-		$url = wp_parse_url( $_SERVER['REQUEST_URI'], PHP_URL_PATH );
-
-		if ( ! empty( $_SERVER['QUERY_STRING'] ) ) {
-			$url .= '?' . $_SERVER['QUERY_STRING'];
-		}
-
-		$request_path = apply_filters( 'wpcom_legacy_redirector_request_path', $url );
+		$request_path = apply_filters( 'wpcom_legacy_redirector_request_path', self::get_current_url_path() );
 
 		if ( $request_path ) {
 			$redirect_uri = self::get_redirect_uri( $request_path );
+
+			if ( !$redirect_uri ) {
+				// trailing slash agnostic support to check if redirect exists by toggling slash based on current URL.
+				$request_path = apply_filters( 'wpcom_legacy_redirector_request_path', self::get_current_url_path( TRUE ) );
+				$redirect_uri = self::get_redirect_uri( $request_path );
```

Comment:
@GaryJones There are a few possibilities.

What about checking if the site’s permalink structure includes a trailing slash? There’s a function to check this user_trailingslashit().

This can then be applied consistently to both get_url_hash() and the request URL being checked through maybe_do_redirect(). You’ll get match that way the first time, with or without the trailing slash, since a site wide standard will be enforced.

The caveat is we’d have to decide if we want to support allowing separate redirects for the both variants (/foo/ and /foo).

Reply:
Correct - lookup what the current permalink has, and change the request URL value to include / not include that, before the `wpcom_legacy_redirector_request_path` filter has run so that clients can still override our logic, but it happens all before the first lookup.

Here, we make the decision to enforce the trailing slash or not, but let the client make the decision if they agree or not.

Reply:
@GaryJones The only challenge with this approach is for existing redirects. As get_url_hash() (post_name and cache key) will potentially differ since we're now applying user_trailingslashit() policy to URL's. It may become out of agreement with any redirects that have been already created if.

For example, /previous-redirect is considered to be /previous-redirect/ because the permalink structure now includes a trailing slash.

Reply:
Also, are opposed to adding another cache item that has the actual redirect destination value? The value would be equal to the return value of get_redirect_uri().

That will eliminate 99% of the current flow as you've outlined for get_redirect_uri() processing. We'd immediately check the cache for an exact match based on the current URL to speed things up.